src/services/WaveformDataUtils.js

```diff
@@ -44,4 +48,136 @@ export default class WaveformDataUtils {
 
     return peaksInstance;
   }
+
+  /**
+   * Add a new segment to Peaks when a new timespan is created
+   * @param {Array} smData - updated structured metadata
+   * @param {Object} peaksInstance - peaks instance for the waveform
+   */
+  insertNewSegment(smData, peaksInstance) {
+    const allSpans = structMetadataUtils.getItemsOfType('span', smData);
+    const segments = peaksInstance.segments.getSegments();
+    const newSpan = allSpans.filter(
+      span => !segments.some(segment => segment.id === span.id)
+    );
+    const { begin, end, label, id } = newSpan[0];
+    peaksInstance.segments.add({
+      startTime: structMetadataUtils.toMs(begin) / 1000,
+      endTime: structMetadataUtils.toMs(end) / 1000,
+      labelText: label,
+      id: id
+    });
+
+    return this.updateSegmentColors(peaksInstance, smData);
```

Comment:
Yes. Removing `updateSegmentColors()` from within that function would be a good idea. And keep it as a separate function called `rebuildPeaks()` or something, which gets called within the reducer after updating the `peaksInstance`. That way we can make sure it is invoked when an update goes through.

The reason for passing `smData` data is to get the order of spans in the structure, so that we can rearrange alternating the colors in segments. In the `updateSegmentColors()` function I'm removing all segments and adding them back with changed colors. So it kind of need the whole structure at this moment. And this is going to be a very expensive action when the structures get very big as you pointed out. I will see whether this can be done without clearing out the segments list. That way I might be able to remove passing `smData` into this function.